betatest/forms.py

```diff
@@ -0,0 +1,205 @@
+from django.utils.text import capfirst
+from djng.forms import NgFormValidationMixin, NgModelForm, NgModelFormMixin
+from zxcvbn import password_strength
+
+from betatest.models import BetaTestApplication
+from userprofile.models import UserProfile
+
+
+# Forms #######################################################################
+
+class BetaTestApplicationForm(NgModelFormMixin, NgFormValidationMixin, NgModelForm):
+    """
+    Application form for beta testers. Creates User, UserProfile, and
+    BetaTestApplication models.
+    """
+    class Meta: #pylint: disable=missing-docstring
+        model = BetaTestApplication
+        exclude = ('user', 'status')
+
+    field_order = (
+        'subdomain',
+        'instance_name',
+        'full_name',
+        'username',
+        'email',
+        'public_contact_email',
+        'password',
+        'password_confirmation',
+        'project_description',
+        'accept_terms',
+        'subscribe_to_updates',
+    )
+
+    full_name = forms.CharField(
+        max_length=255,
+        help_text='Example: Albus Dumbledore',
+    )
+    username = forms.RegexField(
+        regex=r'^[\w.+-]+$',
+        max_length=30,
+        help_text=('This would also be the username of the administrator '
+                   'account on the Open edX instance.'),
+        error_messages={
+            'invalid': ('Usernames may contain only letters, numbers, and '
+                        './+/-/_ characters.'),
+            'unique': 'This username is already taken.',
+        },
+    )
+    email = forms.EmailField(
+        help_text=('This is also your account name, and where we will send '
+                   'important notices.'),
+    )
+    password = forms.CharField(
+        widget=forms.PasswordInput,
+        help_text=('Pick a password for your OpenCraft account. You will be '
+                   'able to use it to login and access your account.'),
+    )
+    password_confirmation = forms.CharField(
+        widget=forms.PasswordInput,
+    )
+    accept_terms = forms.BooleanField(
+        required=True,
+        label='',
+        help_text=('I understand that this is a beta test, that bugs and '
+                   'crashes are expected, and that the instance is provided '
+                   'for free for the duration of the beta-test, without any '
+                   'guarantee.'),
+        error_messages={
+            'required': 'You must accept these terms to register.',
+        },
+    )
+
+    # This field is created automatically from the model field, but the regex
+    # validator is not copied over. We need to define the field manually so
+    # that validation will work client-side. We do this by copying from the
+    # model field (DRY).
+    _subdomain_field = Meta.model._meta.get_field('subdomain')
+    _subdomain_validator = next(v for v in _subdomain_field.validators
+                                if hasattr(v, 'regex'))
+    subdomain = forms.RegexField(
```

Comment:
Do we validate whether the subdomain is already taken?  We should look at both the instances that are currently running and the other beta test applications.  It might also make sense to blacklist certain values (we don't really want anyone to register `www.openedxhosting.com`), but we could do that manually for now.